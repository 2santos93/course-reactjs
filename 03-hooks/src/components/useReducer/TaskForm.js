import React, { forwardRef } from 'react'

export const TaskForm = forwardRef(({onClickHandler}, ref) => {
    return (
        <form>
          <label htmlFor="task" className="form-label">
            <h2>Task</h2>
          </label>
          <input
            ref={ref}
            id="task"
            type="text"
            className="form-control"
          />
          <button
            type="submit"
            className="btn btn-primary mt-3"
            onClick={onClickHandler}
          >
            Submit
          </button>
        </form>
    )
})
