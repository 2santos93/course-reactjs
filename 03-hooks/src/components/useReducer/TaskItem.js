import ACTIONS from "./constant";
import "./style.css";

export const TaskItem = ({ pos, btnClickHandler, task }) => {
  
    const {task:taskName, id, done} = task;

  return (
    <div
      style={{ display: "flex", justifyContent: "space-between" }}
      className="mt-2"
    >
      <h3 className={`${done && "complete"}`}>
        {pos}. {taskName}
      </h3>

      <button
        className="btn btn-success btn-sm"
        onClick={() => btnClickHandler({ type: ACTIONS.UPDATE, payload: id })}
      >
        Toggle
      </button>
      <button
        className="btn btn-danger btn-sm"
        onClick={() => btnClickHandler({ type: ACTIONS.DELETE, payload: id })}
      >
        Delete
      </button>
    </div>
  );
};

TaskItem.displayName='TaskItem';