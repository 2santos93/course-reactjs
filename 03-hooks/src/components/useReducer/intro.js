const initialState = [{
    id: '1',
    toDo: 'homework',
    done: false
}]

const reducer = (state = initialState, action) => {

    if(action?.type === 'add'){
        state = [...state, action.payload];
    }
    return state;
}

let toDoReducer = reducer();

const newToDo = {
    id: '2',
    toDo: 'work',
    done: false
};

const action = {
    type: 'add',
    payload: newToDo
};

toDoReducer = reducer(initialState, action);

console.log(toDoReducer);

