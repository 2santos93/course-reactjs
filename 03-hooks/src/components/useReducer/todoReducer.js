import ACTIONS from './constant';

export const todoReducer = (state, action) => {
    let id;
    switch(action.type){
        case ACTIONS.CREATE:

            return [...state, action.payload];

        case ACTIONS.UPDATE:

            id = action.payload;
            return state.map((task) => (
                (task.id === id)
                    ? {...task, done: !task.done}
                    : task
            ));

        case ACTIONS.DELETE:
            
            id = action.payload;
            return state.filter((task) => task.id !== id);

        default:
            return state;
    }
}
