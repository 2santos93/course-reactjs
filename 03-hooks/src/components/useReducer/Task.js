import React, { useEffect, useReducer, useRef, forwardRef} from "react";

import { todoReducer } from "./todoReducer";
import { TaskItem } from "./TaskItem";
import {TaskForm} from './TaskForm';
import ACTIONS from "./constant";
import {init} from '../../helper/init';

export const Task = forwardRef((props, ref) => {

  const [tasks, dispatch] = useReducer(todoReducer, [], init);
  const inputRef = useRef("");

  useEffect(() => {
    localStorage.setItem("tasks", JSON.stringify(tasks));
    ref.current = tasks.length;
  }, [tasks]);

  const onClickHandler = (e) => {
    e.preventDefault();

    const input = inputRef.current;

    if (!input.value.trim().length > 0) return;

    const task = {
      id: Date.now(),
      task: input.value,
      done: false,
    };

    dispatch({
      type: ACTIONS.CREATE,
      payload: task.id,
    });

    input.value = "";
  };

  const btnClickHandler = (action) => {
    dispatch(action);
  };

  return (
    <div className="row">
      <div className="col-md-3">
        <TaskForm 
            onClickHandler={onClickHandler}
            ref={inputRef}
        />
      </div>
      <div className="offset-md-1 col-md-3">
        {tasks &&
          tasks.map((task, pos) => (
            <TaskItem
              key={task.task}
              pos={pos + 1}
              btnClickHandler={btnClickHandler}
              task={task}
            />
          ))}
      </div>
    </div>
  );
});
