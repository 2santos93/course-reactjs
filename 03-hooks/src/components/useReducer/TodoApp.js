import React, {useRef } from "react";
import { Task } from "./Task";

export const TodoApp = () => {

  const tasksRef = useRef([]);

  return (
    <>
      <h1>ToDo App ({tasksRef.current})</h1>
      <hr />

      <Task
        ref={tasksRef}
      />
    </>
  );
};
