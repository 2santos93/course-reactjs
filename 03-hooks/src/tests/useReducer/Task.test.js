import { Task } from "../../components/useReducer/Task"
import {shallow} from 'enzyme'
import {initialState} from '../Fixtures/todoReducer';
import React from "react";
import {init} from '../../helper/init';

jest.mock('../../helper/init.js');

describe('Task test', () => {

  jest.resetAllMocks();
  const refMock = jest.spyOn(React, 'useRef');
  init.mockReturnValue(initialState);
  const wrapper = shallow(<Task ref={refMock} />)

  test('should render ok', () => {
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('TaskForm').length).toBe(1);
    expect(wrapper.find('TaskItem').length).toBe(1);
  });

})
