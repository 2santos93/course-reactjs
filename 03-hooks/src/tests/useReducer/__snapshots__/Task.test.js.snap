// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task test should render ok 1`] = `
<div
  className="row"
>
  <div
    className="col-md-3"
  >
    <TaskForm
      onClickHandler={[Function]}
    />
  </div>
  <div
    className="offset-md-1 col-md-3"
  >
    <TaskItem
      btnClickHandler={[Function]}
      key="Make DInner"
      pos={1}
      task={
        Object {
          "done": false,
          "id": 1631315572372,
          "task": "Make DInner",
        }
      }
    />
  </div>
</div>
`;
